go/chat/attachment.go

```diff
@@ -1,85 +1,178 @@
 package chat
 
 import (
-	"bytes"
 	"crypto/hmac"
 	"crypto/sha256"
 	"fmt"
 	"io"
-	"time"
+	"path/filepath"
 
 	"github.com/keybase/client/go/chat/s3"
-	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/chat/signencrypt"
 	"github.com/keybase/client/go/logger"
 	"github.com/keybase/client/go/protocol/chat1"
 	"golang.org/x/net/context"
-	"golang.org/x/sync/errgroup"
 )
 
-type ProgressReporter func(bytesCompleted, bytesTotal int)
+type ReadResetter interface {
+	io.Reader
+	Reset() error
+}
+
+type UploadTask struct {
+	S3Params       chat1.S3Params
+	LocalSrc       chat1.LocalSource
+	Plaintext      ReadResetter
+	plaintextHash  []byte
+	S3Signer       s3.Signer
+	ConversationID chat1.ConversationID
+	Progress       ProgressReporter
+}
 
-const minMultiSize = 5 * 1024 * 1024 // can't use Multi API with parts less than 5MB
-const blockSize = 5 * 1024 * 1024    // 5MB is the minimum Multi part size
+func (u *UploadTask) computePlaintextHash() error {
+	plaintextHasher := sha256.New()
+	io.Copy(plaintextHasher, u.Plaintext)
+	u.plaintextHash = plaintextHasher.Sum(nil)
+
+	// reset the stream to the beginning of the file
+	return u.Plaintext.Reset()
 
-// PutS3Result is the success result of calling PutS3.
-type PutS3Result struct {
-	Region   string
-	Endpoint string
-	Bucket   string
-	Path     string
-	Size     int64
 }
 
-// PutS3 uploads the data in Reader r to S3.  It chooses whether to use
-// putSingle or putMultiPipeline based on the size of the object.
-func PutS3(ctx context.Context, log logger.Logger, r io.Reader, size int64, params chat1.S3Params, signer s3.Signer, progress ProgressReporter) (*PutS3Result, error) {
-	region := s3.Region{
-		Name:             params.RegionName,
-		S3Endpoint:       params.RegionEndpoint,
-		S3BucketEndpoint: params.RegionBucketEndpoint,
+func (u *UploadTask) Nonce() signencrypt.Nonce {
+	var n [signencrypt.NonceSize]byte
+	copy(n[:], u.plaintextHash)
+	return &n
+}
```

Comment:
If you want to stick with constant nonce for all attachments that is fine with me.  At one point we discussed using this for the nonce since we had it.

The way it currently stands, we run computePlaintextHash on plaintext in the event of an abort, but I'll add some protection to make sure it never returns an empty nonce.